### ISSUE TITLE
log level by sink

### ISSUE THREAD

**Issue by nmarwen** [gabime/spdlog#79]
I think It can be interesting to have log levels and patterns by sink in addition to the definition by logger.

**Comment by p0fi** [comment]
+1 for this!

It would make it possible to create a logger with multiple sinks (e.g. console output and a rotating file) and write everything in the console, but only the errors in the file!

I'd love to see this feature!

**Comment by gabime** [comment]
I think that in order to keep things consistent and simple, a good approach would be for spdlog to provide a combined logger instead - a logger that holds a list of loggers to forward message to. 
Something like
```c++
auto console = spdlog::console_logger_mt("console");
console->set_level(level::info);

auto file_logger = spdlog::simple_file_logger_mt("filelogger", "filename");
file_logger->set_level(level::debug);
file_logger->set_pattern("some pattern %t %v");

auto combined = spdlog::combined_logger("combined", console, file_logger);
combined->debug("Should not appear on the screen, only in the log file");

```

**Comment by p0fi** [comment]
I can't find a way to combine loggers into one using the current version. Is that something you already implemented or you are going to do this?

As long as this loggers are still thread safe and globally useable like the ones I create right now with this:
```
auto Debug = make_shared<spdlog::logger>("someName", begin(sinks), end(sinks));
spdlog::register_logger(EODebug);
```

it seems to me a rather good approach.

You could set a own pattern and level for this logger, which would basically serve every need I mentioned.

**Comment by gabime** [comment]
I created a [branch](https://github.com/gabime/spdlog/tree/sink_level_support) with support for setting levels on sinks.

Please try and give me feedback on this.

example:
```c++

//first sink with info level
auto sink1 = std::make_shared<spdlog::sinks::stdout_sink_st>();
sink1->set_level(spdlog::level::info);
//second sink with debug level
auto sink2 = std::make_shared<spdlog::sinks::daily_file_sink_st>("logfile", "txt", 23, 59);
sink2->set_level(spdlog::level::debug);

//combined logger with debug level
auto combined = std::make_shared<spdlog::logger>("combined", spdlog::sinks_init_list{ sink1, sink2 });
combined->set_level(spdlog::level::debug);
spdlog::register_logger(combined); //Needed if you want to use spdlog::get("combined") elsewhere

```

The problem here is that a bit confusing - both logger and sinks can now have log levels. 

So if logger has ```info``` level for example, and the user calls ```logger->debug(..)```, the logger will not pass the message to the underlying sinks, even if any of them has ```debug``` level set..

**Comment by p0fi** [comment]
I tried it and its working for most of the cases I think. But there are 2 issues with that:
* it's really a bit confusing, but that doesn't matter actually 

but the second thing is really something which should be changed I think:
* if you have many loggers for many different software parts like I do, you probably want to turn off some log outputs completely. So I sometimes set this logger levels to `off`, but like this you would loose also the error outputs in the file. This error or emergency or some high levels i would like to **always** see in the file, no matter on what level the logger is set actually.

**Comment by gabime** [comment]
Why not just set the level to error/emergency in the logger? If you set it to off, then its off..
I tried with the list of loggers approach. It has its own problems..

**Comment by p0fi** [comment]
Well, thats a point actually!

**Comment by gabime** [comment]
To avoid the level confusion I am thinking about removing the log_level state from the logger entirely - so that only sinks will have actual level

```logger->set_level(..)``` would still exist but all it would do is to set the requested level to its sinks..```

**Comment by p0fi** [comment]
but if I get this right this would mean that if I set the level of a logger (which contains 2 sinks, file and console) would set both sinks to that level, right?

But the (at least for me) intended behavior should be that there is a difference in level for the console and the file. Or should i just avoid setting the level using the logger, and rather set the level using the sink?

But I think that would mean i could not have different levels for my different software parts.

**Comment by nmarwen** [comment]
It looks perfect and I think managing levels by sink it's the perfect solution to remove confusion, so if it's done at logger level this will updates/override sinks levels.

**Comment by gabime** [comment]
@p0fi  right. And this means user would need to set level to sinks to achieve this.

I doesn't feel completely right for me though.. For example if 2 loggers share same sink (typically the stdout_sink) then ```logger->set_level()``` will affect the other logger too (it uses the same sink)

That's why I changed the name to ```logger->set_sinks_level()```..

I think I am going to try the other approach too - a combined logger which forwards its messages to sub loggers. We'll see which approach feels better..

**Comment by p0fi** [comment]
any updates on this here? :)

**Comment by gabime** [comment]
No. Didn't have any time lately

**Comment by mrossi19** [comment]
Any chance this gets merged into master? Need to have everything I log go to simple_file_sink with anything at error or above going to syslog as well.

**Comment by Diom** [comment]
I also would like this feature. I think however the approach is conflating two ideas; filters  & sinks.

_excuse the pseudo code for brevity_

Filters should be seen as standalone items which are associated with sinks. In the same vein sinks are associated with loggers via a sink + filter pairing (`std::vector<std::pair<sink, filter>> sinks;`). The level could be inherited by the logger by selecting the lowest level from the sink filters, but then allowing any level on a logger should be disallowed and the levels should only apply to the sinks (you need at least one after all). Allowing levels to be set at the logger as well would be confusing.

The logic for managing the levels would be simple enough. A `std::array sinkLevel[LEVEL.count];` per logger with a count of sinks at each level. Or iterate over the `sinks` pairs when a new one is added depends on whether you want to limit the cost of adding sinks or the size of the logger.

PS Thanks for such a great logging lib. :)

**Comment by petke** [comment]
I would also have a need for such a feature. 

I'm switching from boost log (which does not seem to work with the soon to be standardised await feature (pplawait.h). I think boost log gets confused because coroutines can start on one thread and then get resumed on another. I'm hoping sdplog works better). 

Anyways in boost.log I have one global logger with many sinks. Anything above "info" level shows up on the console, anything above trace is written to a (huge) debug log file, and anything above error is written to a (small) error file. I'm not sure how to do something similar with spdlog. 

With spdlog, If I understand correctly, I can combine multiple sinks and assign them to the same logger. But all sinks will end up with the same log level. While I'd need them to be different.

Alternatively if I could combine multiple loggers, and these loggers could have different log levels that would solve it. But it seems the function combined_logger (as mentioned above) is not checked in to master.

Thanks /Petke

**Comment by aaost** [comment]
IMO, this would be a pretty important feature to have. I'm coming from log4cxx where each line was tagged with the module that produced it and I'm trying to emulate this with spdlog. So instead of having two loggers named console_logger and file_logger, I have a single logger with two sinks (console/file) attached to each module. While I would like to get detailed debug information in the log file, I don't want to see it in the console. Another use case is sometimes my tests fail on the buildserver and on the buildserver only and I'd like to go through the logs to find out why (but with the way they are set up I can't have anything extraneous output to the console)

**Comment by aggieNick02** [comment]
Another vote for being able to chain two loggers together. Coming from boost.log, the use case @aaost mentioned is easy to support, and was exactly what I wanted/implemented in my app. But then I looked at the tremendous performance hit boost.log brought and it is not acceptable.

So now I'm evaluating spdlog and the first thing I looked for was "how can I get important log messages to go to both console and file, but debug/info type messages to go only to file". With spdlog as is, I have to manually send the message to two different loggers. If it supported creating a logger with subloggers, whose only purpose was to forward every sort of message onto the subloggers, that would be really helpful.

**Comment by Dysl3xik** [comment]
Any updates on this? I am in the middle of adding this to a new project and really like the idea of being able to dynamically adjust where logs go and at which level. For example I want to write everything info and higher to the console and a "normal" textlog. But for debugging (runtime optional) I would want to write all trace and higher to a "debug" textlog + normal logging. 

I realize I can get this to happen by adding 2 loggers, but then I have to "get" both loggers and then pass the log message to both. This isn't incredibly painful but seems wasteful. It would be nice to have 1 logger with these 3 sinks and set the level of each as appropriate i.e. debug is off by default and others are set to normal.

If its going to be a while I can make a custom sink and do this myself, but that makes dynamic level switching a bit more complicated.

Also side note - Thanks for putting the lib together, its incredibly easy to pickup and use, and very lightweight!

**Comment by gabime** [comment]
@Dysl3xik its in the todo list.. Still not sure what the best approach would be

**Comment by facontidavide** [comment]
I noticed that the current logger is already multi-synk
https://github.com/gabime/spdlog/blob/master/include/spdlog/details/logger_impl.h#L18
https://github.com/gabime/spdlog/blob/master/include/spdlog/logger.h#L83

Doesn't this mean that the standard spdlog::logger is already composable?
Can/t we just add 

`std::vector<spdlog::level_t> _sink_levels;`

And check that in the method __spdlog::logger::_sink_it__ ?

**Comment by gabime** [comment]
Will be problematic. For example when sink's level get modified directly or by other logger.

**Comment by petke** [comment]
I probably shouldn't comment as I don't remember how most of this works. Here is my 2 cents anyway: 

Currently log_level is at logger level. I think ideally the log_level belongs in the sink level (so the same logger can have multiple sinks with different log levels). Simply moving the log_level from logger to sink, would brake the old interface. So that probably cant be done. How about having it in both places then? 

Option 1) The two log_level's could be seen as AND filters. The log_level in logger would filter out some messages, and the log_level in sink could filter out some more. So for a message to get through, its log_level would have to be higher than both the logger AND the sink. This might be a bit confusing to users.

Option 2) The log_level in logger is used by default. The user has to call some function to tell the logger to use the sink level log levels instead. Its a bit ugly, but probably less confusing to users.

**Comment by gabime** [comment]
@petke option 1) seems better for me. It is a similar approach to log4j's approach (http://stackoverflow.com/q/1839647)

**Comment by petke** [comment]
Nice. If log4j has support for it, then such an approach shouldnt be too confusing to users of this library either. I'd find it useful, and probably other users too.  Cheers

**Comment by facontidavide** [comment]
The limitation (or not) to this approach (option 1 described by @petke), is that all the sink shared the same format.
This is cool in many cases.

A more flexible option MAYBE is to keep the (1 logger = 1 sink) assumption and to build a __composable_logger__ that actually work as a way to send a log message to multiple independent loggers.

This make more sense in my opinion because otherwise you need something similar to spdlog::get() also for sinks. This would mean that sinks require their own name and loglevel. Not a big deal, but it affects the entire API.

I will work on my separate branch and suggest a PR ;)

**Comment by facontidavide** [comment]
Take a look to this branch and tell me what you think.
This is not ready for PR, it is just an example to show what I mean.

https://github.com/facontidavide/spdlog/tree/composable_logger

**Comment by gabime** [comment]
@facontidavide Interesting approach. Although ```logger_comp```should be of a ```spdlog::logger``` type (inherit from spdlog::logger and implement its virtual functions (like async_logger)) to be able to be registered  (spdlog::get(..), etc ). 

Also in case of async mode, things could get messy since each child logger has its own queue and worker thread, so for example an innocent looking logger with 3 sub loggers would create  3 queues and 3 threads.

**Comment by facontidavide** [comment]
I see.

1) I agree that __logger_comp__ should be registered. I am not so sure that inheriting from  `spdlog::logger` makes sense, since that class requires one or multiple syncs to be passed in the constructor and this is not the semantic I want.

2) Considering this, the log hanging fruit is the other approach, i.e. adding log level to __sinks__.

3) about async... I will open another GH issue ;)

**Comment by facontidavide** [comment]
OK.
in my last commit I modified the composed logger and it now inherits from logger.
The code is much cleaner, indeed.

https://github.com/facontidavide/spdlog/commit/24443fb7e4b95ec7823abc15a17d85710fbc3b23

**Comment by Dysl3xik** [comment]
I think I see a flaw in this current plan IMO. 

If for example you want to always write errors to the console (or any sink for that matter), but then if in debug mode you want to write debug to the same sink. If you do this with composing loggers you will end up with double messages going to the sink, whereas if you just toggle the level of the sink and have one "master" logger you avoid this doubling issue.

I don't think the idea of adjusting sinks logging level is the same as a composite logger, and this approach is less flexible. 

I am a big fan of NLog (.NET logger) and their library is very similar, but they keep sinks and loggers seperate, and use "rules" to tie them together. This allows you to have the behavior I described above where you overlap writing to the same sink, but avoid double output https://github.com/NLog/NLog/wiki/Tutorial#multiple-targets see multiple targets and logger specific routing. Their concept of "final" may be the fix for the doubling issue here.

Anyways I am unsure of the exact best fix here, but I can see these composite loggers causing issues with common sinks and double messages.

**Comment by facontidavide** [comment]
@Dysl3xik  I am not quite sure I follow you.

> If for example you want to always write errors to the console, but then if in debug mode you want to write debug to the same sink. If you do this with composing loggers you will end up with double messages going to the sink

I believe you misunderstood (or I missed your point), because in my current implementation there are no double message, unless __that is what you want to do__. I would never follow the partner you described.

https://github.com/facontidavide/spdlog/blob/composable_logger/example/example_comp.cpp#L22

I __intentionally__ used the three console sinks in my example (for the sake of simplicity) but that is not an error or side effect, it is what I specified.
In a real world use case I would have a logger for the console, a second logger to file and maybe a third to store it into a database (custom sink).

Summarizing: an user that want a specific format for each sink, might prefer the comp_logger, while those user that want exactly the same message delivered to multiple sinks, each with its own log level, should use the second example.

**Comment by Dysl3xik** [comment]
Ok I actually went through you code and demo now, earlier I didn't have studio available and was just going off your comments. 

It looks like you added sink levels & composite logging, but the sink levels are "overridden" by the logger level. Based on your earlier comments I thought you had only added the composite logger. I think adding both is great!

The fist case in your demo is exactly what I would want to avoid - multiple loggers writing the same message to the same sink.

In my use case (maybe a common one?) default logging is warning and up and would go to console + a file. Then if debug is enabled (runtime toggle) I would want debug level and up to go to console, and another special debug file. So normally its console + normal file, while in debug its console (debug level msg) + debug file + regular file. I would like to be able to do this with a single logger so I don't have to write redundant logger->info(x) logger2->info(x) lines.

If I could adjust sinks directly then I can get the desired behavior by adding a debug file sink to logger, and adjusting the console sink level down to debug, the existing normal file sink would be untouched and the top level logger would be untouched.  I think I can get this behavior with your second example actually by just setting logger level to trace always, and then tweaking sinks as I wish.

But using composite logger I think I can get the same behavior too (as you mentioned in your comment above) by having 1 logger for console, 1 for debug file, and 1 for normal file and combining these into a composite logger and adjusting the low level loggers individually. 

So I think both of these can solve my issue, now that I realize you should do 1 sink per logger then concat them, then adjust low level loggers accordingly. Thanks for this - it just didn't click before.

tldr; I think these additions would solve a lot of issues and make logging way more flexible, esp runtime adjustments. 

Hopefully this makes more sense than my last post? I tweaked it a bunch so forgive typos haha.

**Comment by facontidavide** [comment]
I am happy to see that we are on the same page ;)
Anyway, don't forget that __the only reason__ to prefer composite logger is to have a different format on each of the sinks.
otherwise the multisink is more efficient because the formatter is called only once.

> In my use case (maybe a common one?) default logging is warning and up and would go to console + a file. Then if debug is enabled (runtime toggle) I would want debug level and up to go to console, and another special debug file. So normally its console + normal file, while in debug its console (debug level msg) + debug file + regular file. 

Let's see how this would looks like in code:

https://github.com/facontidavide/spdlog/blob/composable_logger/example/example_comp.cpp

As you can see I modified the example to do exactly what you described.

Summarizing: @gabime will you consider a pull request that implement this changes?
It is up to you to decide if one or both these approaches are valid and sufficiently clear for the user.

What do you think?

**Comment by gabime** [comment]
@facontidavide Lets start with adding levels to the sinks.. I think this would answer most of the needs.  Later, there is enough demand ,we might  add the coposite logger too.

**Comment by petke** [comment]
Nice work.

The composite logger supporting different format is a nice and useful feature. A composite logger uses up one thread per "sub-logger" right? And the threads spend some time synchronising those threads? I'd be a bit hesitant to use it for that reason. A light and fast logger library that doesn't use a lot of system resources (like threads, memory and cpu time) is more important to me. So support for different formats of the sinks is nice but not a "must have" for me.

In case anyone is interested in my preferred logging use case: 

I use the console for giving feedback to the application user, while the log files are for me.

I like to have 4 sinks for the same logger: Lets call them console_user_info, file_record_info, file_record_error, and file_record_debug.

console_user_info:
I write to the console to inform the application user of INFO/ERROR messages. This informs the user of the progress, statistics, and any errors of the long running server application.

file_record_info:
I also log those same messages to file, so that I can see afterwards what the user saw. To "fact check"  what he is talking about in the bug reports. This files gives a nice overview of the applications usage.

file_record_debug:
I also write those messages and the DEBUG (+TRACE) messages to a debug log file. Few messages are written to console, while a huge amount is written to the debug logs. I can see in detail what happened afterwards in the debug file to track down bugs. Its too much to just get an overview though.

file_record_error:
Finally I also have a file that just logs the error messages. That way I can just glance at the logger folder once in a while, to see if something unexpected has happened. If an error log file exists and is larger than 0kb, I know to go fishing for gory details in the debug logs.

The format for the log files includes the LINE and FILE of the source code where the the messages where written. That way I can tracking down bugs easier when looking at the logs. The application user looking at the console doesn't really care about that info though, its just noise to them. So its nice having a different format for the user console sink. But exposing that info to the application users, is  not the end of the world.

Cheers

**Comment by facontidavide** [comment]
> A composite logger uses up one thread per "sub-logger" right?

Unfortunately it doesn't, for this reason I opened another another issue https://github.com/gabime/spdlog/issues/275 and a PR https://github.com/gabime/spdlog/pull/278 .

Currently, spdlog use one thread per async logger; therefore this proposal is not worse than what user _need to do__ right now. 
In other words, a composable logger is not better or worse than multiple loggers, just more convenient to use.

> I like to have 4 sinks for the same logger: Lets call them console_user_info, file_record_info, file_record_error, and file_record_debug.

This means that you already have 3 threads, one for each file logger.

**Comment by petke** [comment]
ok, thanks for clearing that up for me.

**Comment by gabime** [comment]
Fixed in PR  #279 (thanks to @facontidavide )
